lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,607 @@
+
+    local max_col, max_row, sum_w, sum_h = {}, {}, 0, 0
+
+    local spacing = (self._private.spacing or 0)
+        + (self._private.border_width or 0)
+
+    local max_h, total_col = 0, self._private.column_count
+
+    for row_idx, row in ipairs(self._private.rows) do
+        -- This can be a sparse table
+        local col_count = get_size(row)
+        for col_idx = 1,col_count  do
+            local wdg = row[col_idx]
+            if wdg then
+                local w, h = base.fit_widget(self, context, wdg, width, height)
+                max_col[col_idx] = math.max(max_col[col_idx] or 0, w)
+                max_row[row_idx] = math.max(max_row[row_idx] or 0, h)
+                max_h = math.max(max_h, h)
+            end
+        end
+
+        -- Rows too can be empty, but the table should not be sparse
+        max_row[row_idx] = max_row[row_idx] or 0
+        total_col = math.max(col_count, total_col)
+
+        height = height - max_h - spacing
+        sum_h = sum_h + max_row[row_idx]
+    end
+
+    assert(self._private.column_count == total_col)
+
+    for i=1, self._private.column_count do
+        max_col[i] = max_col[i] or 0
+        sum_w = sum_w + max_col[i]
+    end
+
+    return max_col, max_row, sum_w, sum_h
+end
+
+local function get_ajusted_cell_sizes(self, context, width, height)
+    local max_col, max_row, sw, sh = get_cell_sizes(self, context, width, height)
+
+    local spacing = (self._private.spacing or 0)
+        + (self._private.border_width or 0)
+
+    if self._private.fill_space then
+        if sw + #max_col*spacing < width then
+            local to_add = (width - (sw + #max_col*spacing)) / #max_col
+            for i=1, #max_col do
+                max_col[i] = max_col[i] + to_add
+            end
+        end
+
+        if sh + #max_row*spacing < height then
+            local to_add = (height - (sh + #max_row*spacing)) / #max_row
+            for i=1, #max_row do
+                max_row[i] = max_row[i] + to_add
+            end
+        end
+        sw, sh = width - spacing*#max_col, height - spacing*#max_row
+    end
+
+    return max_col, max_row, sw, sh
+end
+
+-- The grid layout.
+-- @param context The context in which we are drawn.
+-- @param width The available width.
+-- @param height The available height.
+function grid:layout(context, width, height)
+    local max_col, max_row = get_ajusted_cell_sizes(self, context, width, height)
+
+    local result = {}
+
+    local spacing = (self._private.spacing or 0)
+        + (self._private.border_width or 0)
+
+    local y = 0
+
+    for row_idx = 1, #max_row do -- Vertical
+        local x = 0
+        for col_idx = 1, #max_col do -- Horizontal
+            local v = self._private.rows[row_idx][col_idx]
+
+            if v then
+                table.insert(result, base.place_widget_at(
+                    v, x, y, max_col[col_idx], max_row[row_idx]
+                ))
+                x = x + max_col[col_idx] + spacing
```

Comment:
Shouldn't this addition be outside of the `if v then`? Otherwise there is e.g. no spacing around empty cells.